provd/internal/services/accessibility/accessibility_test.go

```diff
@@ -0,0 +1,169 @@
+			}
+			if tc.wmSettingsError {
+				opts = append(opts, accessibility.WithWMSettings(&gSettingsSubsetMock{isWritableError: true}))
+			}
+			if tc.keyboardSettingsError {
+				opts = append(opts, accessibility.WithKeyboardSettings(&gSettingsSubsetMock{isWritableError: true}))
+			}
+
+			client, err := accessibility.New(opts...)
+
+			if tc.wantErr {
+				require.Error(t, err, "New should return an error")
+				require.Empty(t, client, "New should return a nil response")
+				return
+			}
+
+			require.NoError(t, err, "New should not return an error")
+			require.NotNil(t, client, "New should return a non-nil response")
+		})
+	}
+}
+
+// newAccessibilityClient creates a new accessibility client fo			if tc.wantErr {r testing, with a temp unix socket.
+func newAccessibilityClient(t *testing.T, opts ...accessibility.Option) pb.AccessibilityServiceClient {
+	t.Helper()
+
+	// socket path is limited in length.
+	tmpDir, err := os.MkdirTemp("", "hello-socket-dir")
+	require.NoError(t, err, "Setup: could not setup temporary socket dir path")
+	t.Cleanup(func() { _ = os.RemoveAll(tmpDir) })
+	socketPath := filepath.Join(tmpDir, "provd.sock")
+
+	lis, err := net.Listen("unix", socketPath)
+	require.NoError(t, err, "Setup: could not create unix socket")
+
+	service, err := accessibility.New(opts...)
+
+	if err != nil {
+		t.Fatalf("Setup: could not create a11y service: %v", err)
+	}
+
+	grpcServer := grpc.NewServer()
+	pb.RegisterAccessibilityServiceServer(grpcServer, service)
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		_ = grpcServer.Serve(lis)
+	}()
+	t.Cleanup(func() {
+		grpcServer.Stop()
+		<-done
+	})
+
+	conn, err := grpc.Dial("unix://"+socketPath, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	require.NoError(t, err, "Setup: Could not connect to GRPC server")
+	t.Cleanup(func() { _ = conn.Close() })
+
+	return pb.NewAccessibilityServiceClient(conn)
+}
+
+func TestMain(m *testing.M) {
+	testutils.InstallUpdateFlag()
+	flag.Parse()
+
+	m.Run()
+}
+
+type gSettingsSubsetMock struct {
+
+	// Error flags for each method
+	isWritableError bool
+	setBooleanError bool
+	getBooleanError bool
+	setDoubleError  bool
+	getDoubleError  bool
+
+	// Current values
+	currentBool   bool
+	currentDouble float64
+}
+
+func (g gSettingsSubsetMock) IsWritable(key string) bool {
+	return !g.isWritableError
+}
+
+func (g *gSettingsSubsetMock) SetBoolean(key string, value bool) bool {
+	if g.setBooleanError {
+		return false
+	}
+	g.currentBool = value
+	return true
+}
+
+func (g gSettingsSubsetMock) GetBoolean(key string) bool {
+	if g.getBooleanError {
+		return false
+	}
+	return g.currentBool
+}
+
+func (g *gSettingsSubsetMock) SetDouble(key string, value float64) bool {
+	if g.setDoubleError {
+		return false
+	}
+	g.currentDouble = value
+	return true
+}
+
+func (g gSettingsSubsetMock) GetDouble(key string) float64 {
+	if g.getDoubleError {
+		return 0
```

Comment:
My understanding of the underlying func this calls is that is will just return the default value if somthing goes wrong:
```
func (this0 *Settings) GetDouble(key0 string) float64 {
	var this1 *C.GSettings
	var key1 *C.char
	if this0 != nil {
		this1 = (*C.GSettings)(this0.InheritedFromGSettings())
	}
	key1 = _GoStringToGString(key0)
	defer C.free(unsafe.Pointer(key1))
	ret1 := C.g_settings_get_double(this1, key1)
	var ret2 float64
```
It doesn't return a success bool so we can't tell if there was a failure?

Reply:
Exactly! That was my point: the test is lying telling that we error out :) So the idea is really to set to some value and check the return value. If there are 2 values, try both :) but don’t tell there is an error we can trigger (because we can from the mock, but not in reality).

Consequently, the mock should not allow to fail either to reflect the real behaviour.